rpc/subscriptions_test.go

```diff
@@ -326,7 +330,533 @@ func TestSubscribeEvents(t *testing.T) {
+	require.NoError(t, err)
+	require.Equal(t, want, string(block0Got))
+
+	// Simulate a new block
+	syncer.newHeads.Send(testHeader(t))
+
+	// Check new block content
+	_, newBlockGot, err := conn.Read(ctx)
+	require.NoError(t, err)
+	require.Equal(t, newHeadsResponse(id), string(newBlockGot))
+}
+
+func TestMultipleSubscribeNewHeadsAndUnsubscribe(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	t.Cleanup(mockCtrl.Finish)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	mockChain := mocks.NewMockReader(mockCtrl)
+	syncer := newFakeSyncer()
+	handler, server := setupRPC(t, ctx, mockChain, syncer)
+
+	mockChain.EXPECT().HeadsHeader().Return(&core.Header{}, nil).Times(2)
+
+	ws := jsonrpc.NewWebsocket(server, utils.NewNopZapLogger())
+	httpSrv := httptest.NewServer(ws)
+
+	conn1, _, err := websocket.Dial(ctx, httpSrv.URL, nil) //nolint:bodyclose
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, conn1.Close(websocket.StatusNormalClosure, ""))
+	})
+
+	conn2, _, err := websocket.Dial(ctx, httpSrv.URL, nil) //nolint:bodyclose
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, conn2.Close(websocket.StatusNormalClosure, ""))
+	})
+
+	firstID := uint64(1)
+	secondID := uint64(2)
+
+	handler.WithIDGen(func() uint64 { return firstID })
+	firstGot := sendWsMessage(t, ctx, conn1, subMsg("starknet_subscribeNewHeads"))
+	require.NoError(t, err)
+	require.Equal(t, subResp(firstID), firstGot)
+
+	handler.WithIDGen(func() uint64 { return secondID })
+	secondGot := sendWsMessage(t, ctx, conn2, subMsg("starknet_subscribeNewHeads"))
+	require.NoError(t, err)
+	require.Equal(t, subResp(secondID), secondGot)
+
+	// Ignore the first mock header
+	_, _, err = conn1.Read(ctx)
+	require.NoError(t, err)
+	_, _, err = conn2.Read(ctx)
+	require.NoError(t, err)
+
+	// Simulate a new block
+	syncer.newHeads.Send(testHeader(t))
+
+	// Receive a block header.
+	_, firstHeaderGot, err := conn1.Read(ctx)
+	require.NoError(t, err)
+	require.Equal(t, newHeadsResponse(firstID), string(firstHeaderGot))
+
+	_, secondHeaderGot, err := conn2.Read(ctx)
+	require.NoError(t, err)
+	require.Equal(t, newHeadsResponse(secondID), string(secondHeaderGot))
+
+	// Unsubscribe
+	unsubMsg := `{"jsonrpc":"2.0","id":1,"method":"juno_unsubscribe","params":[%d]}`
+	require.NoError(t, conn1.Write(ctx, websocket.MessageBinary, []byte(fmt.Sprintf(unsubMsg, firstID))))
+	require.NoError(t, conn2.Write(ctx, websocket.MessageBinary, []byte(fmt.Sprintf(unsubMsg, secondID))))
+}
+
+func TestSubscriptionReorg(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	mockCtrl := gomock.NewController(t)
+	t.Cleanup(mockCtrl.Finish)
+
+	mockChain := mocks.NewMockReader(mockCtrl)
+	syncer := newFakeSyncer()
+	handler, server := setupRPC(t, ctx, mockChain, syncer)
+
+	testCases := []struct {
+		name            string
+		subscribeMethod string
+		ignoreFirst     bool
```

Comment:
I am a little confused about this. Could you explain when you would want to `ignoreFirst` and when you wouldn't?

Reply:
In `subscribeNewHeads`, the server will also send the latest header stored in the db (done via `sendHistoricalHeaders()`), while processing the new headers sent through the feed. Because this test case wants to test if the headers sent through the feed is working or not, that's why I had an `ignoreFirst` flag.

In `subscribeEvents`, this doesn't happen if I understand correctly.

Reply:
> In subscribeEvents, this doesn't happen if I understand correctly.

Well, this depends on the filter. If the filter is empty, then all the historical and/or current block/s events are streamed to the client; if it is not nil, only the events that match the filter are. In this case, since the header is empty, which makes it EventBloom nil indicating that the block has no events, no such first message is sent.

I think it is fine as is, no change is required.